### DISCUSSION TITLE
API to create/update/delete notebooks in a server session

### DISCUSSION THREAD

**Discussion by fonsp** [JuliaPluto/PlutoSliderServer.jl#4] [Ideas]
From @pankgeorg: It would be cool if we had a way to change the set of notebooks that is running on the server, while the server is running.

This would really help with the deploy times of sites like computationalthinking.mit.edu, which currently need to reboot all notebooks when one file changes.

**Comment by fonsp** [comment]
### Syncing with a git branch
The SliderServer launches on a directory, and we could just add one command (HTTP POST endpoint?) to rescan the directory, boot up notebooks that were not there before, shut down notebooks that got deleted, and update changed notebook files using https://github.com/fonsp/Pluto.jl/pull/1029

That PR uses the enter-exit-update idea from d3 to update only the necessary parts of the notebook. My proposal is essentially to also do that on the folder level.

**Reply by fonsp** [reply]
We could also add an endpoint `POST /git_pull_and_rescan`, you could directly use that link as the github webhook.

**Reply by pankgeorg** [reply]
I love the idea of launching on directory and rescan it for changes, primarily because the API access needs secret management, which can be tricky.

EDIT: [We would still need to somehow remotely update the files on the remote server; maybe doing a `git pull` every once in a while on the folder that's being watched. === `POST /git_pull_and_rescan` with some security]

**Comment by fonsp** [comment]
### CRUD API

We add simple REST endpoints to Create, Read, Update and Delete notebooks from the session. This is simple to implement, but i guess that it would need accompanying scripts to support important use cases (please edit):
- Sync to github branch
- ...

**Reply by pankgeorg** [reply]
Since notebooks are hash referenced the UPDATE is the same as CREATE (new hash)
The read endpoint doesn't make much sense: (read state? bind state? notebook file? session?)
It could be more of a `list` of active sessions or a `GET` that returns if `hash` is running.
`delete` stops a notebook from running
`post` (with a notebook in the body) starts a notebook.

We could also provide an admin panel for these stuff, that will use these endpoints for the standard case.
